File: Cpp/707.设计链表.cpp
/*
 * @lc app=leetcode.cn id=707 lang=cpp
 *
 * [707] 设计链表
 */

#include <iostream>
#include <string>
#include <queue>
#include <unordered_set>

using namespace std;

// @lc code=start
struct ListNode {
    int val;
    ListNode *prev, *next;
    ListNode(int val) : val(val), prev(nullptr), next(nullptr) {}
};

class MyLinkedList {
private:
    int size;
    ListNode *head;
    ListNode *tail;
public:
    MyLinkedList() {
        this->size = 0;
        this->head = new ListNode(0);
        this->tail = new ListNode(0);
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }
        ListNode *cur;
        if (index + 1 < size - index) {
            cur = head;
            for (int i = 0; i <= index; ++i) {
                cur = cur->next;
            }
        } else {
            cur = tail;
            for (int i = 0; i < size - index; ++i) {
                cur = cur->prev;
            }
        }
        return cur->val;
    }
    
    void addAtHead(int val) {
        addAtIndex(0, val);
    }
    
    void addAtTail(int val) {
        addAtIndex(size, val);
    }
    
    void addAtIndex(int index, int val) {
        if (index > size) {
            return;
        }
        index = max(0, index);
        ListNode *pred, *succ;
        if (index < size - index) {
            pred = head;
            for (int i = 0; i < index; ++i) {
                pred = pred->next;
            }
            succ = pred->next;
        } else {
            succ = tail;
            for (int i = 0; i < size - index; ++i) {
                succ = succ->prev;
            }
            pred = succ->prev;
        }
        ++size;
        ListNode *node = new ListNode(val);
        node->prev = pred;
        node->next = succ;
        pred->next = node;
        succ->prev = node;
    }
    
    void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        ListNode *pred, *succ;
        if (index < size - index) {
            pred = head;
            for (int i = 0; i < index; ++i) {
                pred = pred->next;
            }
            succ = pred->next->next;
        } else {
            succ = tail;
            for (int i = 0; i < size - index - 1; i++) {
                succ = succ->prev;
            }
            pred = succ->prev->prev;
        }
        --size;
        ListNode *node = pred->next;
        pred->next = succ;
        succ->prev = pred;
        delete node;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
        
    /**
     MyLinkedList linkedList = new MyLinkedList();
     linkedList.addAtHead(1);
     linkedList.addAtTail(3);
     linkedList.addAtIndex(1,2);   //链表变为1-> 2-> 3
     linkedList.get(1);            //返回2
     linkedList.deleteAtIndex(1);  //现在链表是1-> 3
     linkedList.get(1);            //返回3
     */
    MyLinkedList* linkedList = new MyLinkedList();
    linkedList->addAtHead(1);
    linkedList->addAtTail(3);
    linkedList->addAtIndex(1, 2);
    assert(2 == linkedList->get(1));
    linkedList->deleteAtIndex(1);
    assert(3 == linkedList->get(1));

    cout << "OK~" << endl;
    
    return 0;
}

File: Swift/707.设计链表.swift
/*
 * @lc app=leetcode.cn id=707 lang=swift
 *
 * [707] 设计链表
 */

// @lc code=start

class MyLinkedList {
    
    private class Node {
        let value: Int
        var prev: Node?
        var next: Node?
        
        init(value: Int, prev: Node? = nil, next: Node? = nil) {
            self.value = value
            self.prev = prev
            self.next = next
        }
    }
    
    private var size = 0
    private var head = Node(value: 0)
    private var tail = Node(value: 0)

    init() {
        head.next = tail
        tail.prev = head
    }
    
    func get(_ index: Int) -> Int {
        guard (0..<size).contains(index) else { return -1 }
        
        var node: Node?
        if index + 1 < size - index {
            node = head;
            for _ in 0...index {
                node = node?.next
            }
        } else {
            node = tail
            for _ in 0..<(size - index) {
                node = node?.prev
            }
        }
        return node?.value ?? -1
    }
    
    func addAtHead(_ val: Int) {
        addAtIndex(0, val)
    }
    
    func addAtTail(_ val: Int) {
        addAtIndex(size, val)
    }
    
    func addAtIndex(_ index: Int, _ val: Int) {
        guard index <= size else { return }
        
        if (index > size) {
            return;
        }
        var index = max(0, index)
        var prev: Node?
        var succ: Node?
        if index < size - index {
            prev = head
            for _ in 0..<index {
                prev = prev?.next
            }
            succ = prev?.next
        } else {
            succ = tail
            for _ in 0..<(size - index) {
                succ = succ?.prev
            }
            prev = succ?.prev
        }
        size += 1
        let node = Node(value: val)
        node.prev = prev;
        node.next = succ
        prev?.next = node
        succ?.prev = node
    }
    
    func deleteAtIndex(_ index: Int) {
        guard (0..<size).contains(index) else { return }
        
        var prev: Node?
        var succ: Node?
        if index < size - index {
            prev = head
            for _ in 0..<index {
                prev = prev?.next
            }
            succ = prev?.next?.next
        } else {
            succ = tail
            for _ in 0..<(size - index - 1) {
                succ = succ?.prev
            }
            prev = succ?.prev?.prev
        }
        size -= 1
        prev?.next = succ
        succ?.prev = prev
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * let obj = MyLinkedList()
 * let ret_1: Int = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index, val)
 * obj.deleteAtIndex(index)
 */
// @lc code=end

func main() {
    /**
     MyLinkedList linkedList = new MyLinkedList();
     linkedList.addAtHead(1);
     linkedList.addAtTail(3);
     linkedList.addAtIndex(1,2);   //链表变为1-> 2-> 3
     linkedList.get(1);            //返回2
     linkedList.deleteAtIndex(1);  //现在链表是1-> 3
     linkedList.get(1);            //返回3
     */
    let linkedList = MyLinkedList()
    linkedList.addAtHead(1)
    linkedList.addAtTail(3)
    linkedList.addAtIndex(1, 2)
    assert(2 == linkedList.get(1))
    linkedList.deleteAtIndex(1)
    assert(3 == linkedList.get(1))
}
